modules/fundamental/src/analysis/service/mod.rs

```diff
@@ -0,0 +1,564 @@
+        &self,
+        query: Query,
+        paginated: Paginated,
+        tx: TX,
+    ) -> Result<PaginatedResults<AdvisoryGraphSummary>, Error> {
+        let connection = self.db.connection(&tx);
+        let mut g: Graph<PackageNode, Relationship, petgraph::Directed> = Graph::new();
+
+        // TODO: convert this to 'sea_orm dialect'
+        let sql = format!(
+            r#"
+                WITH filtered_sbom AS (
+                    SELECT DISTINCT ON (document_id) sbom_id
+                    FROM sbom
+                    WHERE sbom_id IN (
+                        SELECT DISTINCT sbom_id
+                        FROM sbom_node
+                        WHERE name ILIKE '%{}%'
+                    )
+                    ORDER BY document_id, published DESC
+                )
+                SELECT
+                    sbom.document_id,
+                    sbom.sbom_id,
+                    sbom.published::text,
+                    get_purl(t1.qualified_purl_id) AS left_qualified_purl,
+                    package_relates_to_package.relationship,
+                    get_purl(t2.qualified_purl_id) AS right_qualified_purl,
+                    product.name AS product_name,
+                    product_version.version AS product_version
+                FROM
+                    sbom
+                LEFT JOIN
+                    product_version ON sbom.sbom_id = product_version.sbom_id
+                LEFT JOIN
+                    product ON product_version.product_id = product.id
+                LEFT JOIN
+                    package_relates_to_package ON sbom.sbom_id = package_relates_to_package.sbom_id
+                LEFT JOIN
+                    sbom_package_purl_ref t1 ON sbom.sbom_id = t1.sbom_id AND t1.node_id = package_relates_to_package.left_node_id
+                LEFT JOIN
+                    sbom_package_purl_ref t2 ON sbom.sbom_id = t2.sbom_id AND t2.node_id = package_relates_to_package.right_node_id
+                WHERE
+                    package_relates_to_package.relationship IN (0, 8, 14)
+                    AND sbom.sbom_id IN (SELECT sbom_id FROM filtered_sbom);
+             "#,
+            query.q.as_str()
+        );
+
+        let relationship_results = connection
+            .query_all(Statement::from_string(
+                connection.get_database_backend(),
+                sql,
+            ))
+            .await?;
+
+        // keep track of nodes / indices for efficient searching
+        let mut nodes: HashMap<String, NodeIndex> = HashMap::new();
+
+        // load package relationships into graph
+        for row in relationship_results {
+            let sbom_published = row
+                .try_get("", "published")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
+            let document_id = row
+                .try_get("", "document_id")
+                .unwrap_or("NOVALUE".to_string()); // TODO: this is not right
```

Comment:
Independent of it's use here, it might be better to use: `.unwrap_or_else(||"NOVALUE".to_string())`.

The difference is that is will only allocate the string when it's really required. Where with `unwrap_or`, it would allocate that string passing it as an argument.